dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java

```diff
@@ -195,17 +180,24 @@ public BooleanFormula visitDifference(Relation rel, Relation r1, Relation r2) {
         @Override
         public BooleanFormula visitComposition(Relation rel, Relation r1, Relation r2) {
             BooleanFormula enc = bmgr.makeTrue();
-            TupleSet r1Set = r1.getEncodeTupleSet();
-            TupleSet r2Set = r2.getEncodeTupleSet();
-            TupleSet minSet = rel.getMinTupleSet();
-            for (Tuple tuple : rel.getEncodeTupleSet()) {
+            final RelationAnalysis.Knowledge k = ra.getKnowledge(rel);
+            final RelationAnalysis.Knowledge k1 = ra.getKnowledge(r1);
+            final RelationAnalysis.Knowledge k2 = ra.getKnowledge(r2);
+            for (Tuple tuple : encodeSets.get(rel)) {
                 BooleanFormula expr = bmgr.makeFalse();
-                if (minSet.contains(tuple)) {
+                if (k.getMustSet().contains(tuple)) {
                     expr = execution(tuple);
                 } else {
-                    for (Tuple t1 : r1Set.getByFirst(tuple.getFirst())) {
-                        Tuple t2 = new Tuple(t1.getSecond(), tuple.getSecond());
-                        if (r2Set.contains(t2)) {
+                    Event x = tuple.getFirst();
+                    Event z = tuple.getSecond();
+                    Set<Tuple> a1 = Sets.union(encodeSets.get(r1), k1.getMustSet());
+                    Set<Tuple> a2 = Sets.union(encodeSets.get(r2), k2.getMustSet());
```

Comment:
These sets can be constructed outside of the loop (the view construction is cheap, but still).